server/models/Backlink.ts

```diff
@@ -37,6 +37,32 @@ class Backlink extends IdModel<
   @ForeignKey(() => Document)
   @Column(DataType.UUID)
   reverseDocumentId: string;
+
+  public static async findSourceDocumentIdsForUser(
+    documentId: string,
+    user: User
+  ) {
+    const backlinks = await this.findAll({
+      attributes: ["reverseDocumentId"],
+      where: {
+        documentId,
+      },
+    });
+
+    const documents = await Document.findByIds(
+      backlinks.map((backlink) => backlink.reverseDocumentId),
+      { userId: user.id }
+    );
+
+    return documents
+      .filter(
```

Comment:
Did you find this filter was required? I would not expect anything to get filtered here 🤔

Reply:
I'm afraid, yes. The membership queries use left outer join (`required: false` [here](https://github.com/outline/outline/blob/main/server/models/Document.ts#L192) & [here](https://github.com/outline/outline/blob/main/server/models/Document.ts#L196) ) -  it'd return all the documents requested. We need to get only the source docs that the user has access to.

Reply:
An example where the filter is necessary.

UserA: owns DocA and DocS.
UserB: owns DocB and has access to DocS.

Consider DocS has backlinks to DocA & DocB.
For UserA, we should list only DocA (likewise DocB for userB).

Reply:
Okay, makes sense – but shouldn't we move this filtering into `findByIds` otherwise it would be easy to misuse that method in the future

Reply:
True - missed it 🤦‍♂️
Filter when "userId" is present.